src/main/java/com/exactpro/th2/crawler/Crawler.java

```diff
@@ -52,205 +53,233 @@
+                EventID startId = null;
 
-                } else if (MESSAGES.equals(interval.getCrawlerType())) {
-                    report = sendMessages(crawlerId, info, batchSize);
-                } else {
-                    throw new ConfigurationException("Type must be either EVENTS or MESSAGES");
+                if (lastProcessedEvent != null) {
+                    if (!fetchIntervalReport.processFromStart) {
+                        startId = EventUtils.toEventID(lastProcessedEvent.getId());
+                    }
                 }
 
-                restartPod = report.action == CrawlerAction.STOP;
+                sendingReport = sendEvents(new EventsInfo(interval, info, startId,
+                        interval.getStartTime(), interval.getEndTime()));
 
-                if (report.action == CrawlerAction.NONE) {
-                    interval = intervalsWorker.setIntervalProcessed(interval, true);
+            } else if (MESSAGES.equals(interval.getCrawlerType())) {
+                Map<CrawlerUtils.AliasAndDirection, RecoveryState.InnerMessage> lastProcessedMessages;
+                RecoveryState state = RecoveryState.getStateFromJson(interval.getRecoveryState());
 
-                    RecoveryState previousState = interval.getRecoveryState();
+                lastProcessedMessages = state == null ? null : state.getLastProcessedMessages();
 
-                    if (EVENTS.equals(interval.getCrawlerType())) {
-                        RecoveryState state = new RecoveryState(
-                                null,
-                                previousState.getLastProcessedMessages(),
-                                numberOfEvents,
-                                previousState.getLastNumberOfMessages());
+                MessageID startId = null;
 
-                        interval = intervalsWorker.updateRecoveryState(interval, state);
-                    }
+                if (lastProcessedMessages != null) { // TODO: check if I take the correct startId here
+                    if (!fetchIntervalReport.processFromStart) {
+                        RecoveryState.InnerMessage latestMessage = lastProcessedMessages.values().stream()
+                                .max(Comparator.comparing(RecoveryState.InnerMessage::getTimestamp))
+                                .orElse(null);
 
-                    numberOfEvents = 0L;
-                    numberOfMessages = 0L;
+                        if (latestMessage != null) {
+                            startId = MessageID.newBuilder().setSequence(latestMessage.getSequence())
+                                    .setConnectionId(ConnectionID.newBuilder().setSessionAlias(latestMessage.getSessionAlias()).build())
+                                    .setDirection(com.exactpro.th2.common.grpc.Direction.valueOf(latestMessage.getDirection().toString()))
+                                    .build();
+                        }
```

Comment:
Why do we select only a single message? We should use all messages from the recovery state and put them as resumed into request to the data-provider